CFP: Pod MAC Address Specification (#29360)
cilium-team/cilium: Documentation/network/pod-mac-address.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _pod_mac_address:

************************************
Use a Specific MAC Address for a Pod
************************************

Some applications bind software licenses to network interface MAC addresses.
Cilium provides the ability to specific MAC addresses for pods
at deploy time instead of letting the operating system allocate them.

Configuring the address
#######################

Cilium will configure the MAC address for the primary interface inside a

> Pod if you specify the MAC address in the ``cni.cilium.io/mac-address`` before
> deploying the Pod. This MAC address is isolated to the container so it will

Reviewer: nit: 
```suggestion
Pod if you specify the MAC address in the ``cni.cilium.io/mac-address`` annotation before
deploying the Pod. This MAC address is isolated to the container so it will
```

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _pod_mac_address:

************************************
Use a Specific MAC Address for a Pod
************************************

> 

Reviewer: nit:
```suggestion
```

Some applications bind software licenses to network interface MAC addresses.

> Typically, the solution is very complex and requires things like Docker within
> the cluster.
> 

Reviewer: nit, I don't think that this sentence provides much clarity for the motivation of the feature. I think that the text would be just as clear without this.

```suggestion
```

> Therefore, cilium provides the ability to specific MAC addresses
> for pods instead of letting the operating system allocate them.

Reviewer: ```suggestion
Cilium provides the ability to specific MAC addresses for pods
at deploy time instead of letting the operating system allocate them.
```

> 
> 
> 

Reviewer: nit: extraneous whitespace
```suggestion
```

> Specific MAC Address
> #####################

Reviewer: ```suggestion
Configuring the address
#######################
```

> The MAC address configured by the annotation described here will be visible
> from within the container on the eth0 interface. Since it is isolated to the
> container it will not collide with any other MAC addresses assigned to other
> pods on the same node.

Reviewer: ```suggestion
Cilium will configure the MAC address for the primary interface inside a
Pod if you specify the MAC address in the ``cni.cilium.io/mac-address`` before
deploying the Pod. This MAC address is isolated to the container so it will
not collide with any other MAC addresses assigned to other Pods on the same
node. The MAC address must be specified *before* deploying the Pod.
```

Reviewer: Minor nit here: The naming of `eth0` for the primary interface is distribution-specific for the container image, so we should prefer `primary interface` and only use `eth0` as an example.

> 

Reviewer: nit:
```suggestion
```

Annotate the pod with ``cni.cilium.io/mac-address`` set to the desired MAC address.

> After adding the annotation to the pod, it appears like so.

Reviewer: ```suggestion
For example:
```

.. code-block:: yaml

    apiVersion: v1
    kind: Pod
    metadata:
      annotations:
        cni.cilium.io/mac-address: e2:9c:30:38:52:61
      labels:
        app: busybox
      name: busybox
      namespace: default

> After the pod runs successfully, we can query the pod's eth0 interface.

Reviewer: If the user needs to do something, we should instruct them to do it in simple terms. We shouldn't assume that they do something via wording like "after the pod runs successfully".

```suggestion
Deploy the Pod. Cilium will configure the mac address to the first interface in the Pod automatically.
```

Author: Updated.

Check whether its mac address is the specified mac address.

.. code-block:: shell-session

    $ kubectl exec -it busybox -- ip addr
    1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1000
        link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
        inet 127.0.0.1/8 scope host lo
           valid_lft forever preferred_lft forever
        inet6 ::1/128 scope host
           valid_lft forever preferred_lft forever
    7: eth0@if8: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue qlen 1000
        link/ether e2:9c:30:38:52:61 brd ff:ff:ff:ff:ff:ff
        inet 10.244.2.195/32 scope global eth0
           valid_lft forever preferred_lft forever
        inet6 fe80::e46d:f4ff:fe4d:ebca/64 scope link
           valid_lft forever preferred_lft forever

.. warning::
    The MAC address must be specified in this annotation during Pod creation,

>     and Cilium will not respect changes to the annotation after Pod creation.

Reviewer: I think that this is critical information, and should be worked into the text directly rather than adding a warning callout at the end of the document.

> .. warning::
>     The MAC address must be specified in this annotation during Pod creation,
>     and Cilium will not respect changes to the annotation after Pod creation.

Reviewer: ```suggestion
```
